File: Baekjoon/Java/숫자판 점프.java
import java.io.*;
import java.util.*;


public class Main {
    static List<String> list = new ArrayList<>();
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        String[][] boards = new String[5][5];
        for (int i = 0; i < 5; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 5; j++) {
                boards[i][j] = st.nextToken();
            }
        }

        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                dfs(i, j, boards[i][j], boards, 0);
            }
        }
        System.out.println(list.size());
    }

    public static void dfs (int x, int y, String number, String[][] boards, int cnt) {
        if (cnt == 5){
            if (!list.contains(number)) {
                list.add(number);
            } return;
        }

        for (int i = 0; i < 4; i++) {
            int rx = x + dx[i];
            int ry = y + dy[i];
            if ((0 <= rx && rx < 5) && (0 <= ry && ry < 5)) {
                dfs(rx, ry, number + boards[rx][ry], boards, cnt + 1);
            }
        }
    }

}

File: Baekjoon/Python/숫자판 점프.py
import sys
input = sys.stdin.readline

def dfs(x,y,number):
    if len(number) == 6:
        result.append(number)
        return
    dir = [(0,1),(1,0),(0,-1),(-1,0)]
    for dx, dy in dir:
        rx = x + dx
        ry = y + dy
        if 0<=rx<5 and 0<=ry<5:
            dfs(rx,ry,number+str(boards[rx][ry]))

boards = []
result = []
for i in range(5):
    boards.append(list(map(int,input().split())))

for i in range(5):
    for j in range(5):
        dfs(i,j,str(boards[i][j]))

print(len(set(result)))



